lib/ts/recipe/totp/components/themes/mfa/totpCodeForm.tsx

```diff
@@ -0,0 +1,100 @@
+import FormBase from "../../../../emailpassword/components/library/formBase";
+import { totpCodeValidate } from "../../../utils";
+
+import type { TOTPMFACommonProps } from "../../../types";
+import type { RecipeInterface } from "supertokens-web-js/recipe/totp";
+
+export const CodeForm = withOverride(
+    "TOTPCodeForm",
+    function TOTPCodeForm(
+        props: TOTPMFACommonProps & {
+            onSuccess: () => void;
+            clearError: () => void;
+            onError: (err: string) => void;
+            footer?: JSX.Element;
+        }
+    ): JSX.Element {
+        const userContext = useUserContext();
+
+        return (
+            <React.Fragment>
+                <FormBase
+                    formDataSupertokens="totp-mfa codeForm"
+                    clearError={props.clearError}
+                    onError={props.onError}
+                    formFields={[
+                        {
+                            id: "totp",
+                            label: "TOTP_CODE_INPUT_LABEL",
+                            autofocus: true,
+                            optional: false,
+                            clearOnSubmit: true,
+                            autoComplete: "one-time-code",
+                            placeholder: "",
+                            validate: totpCodeValidate,
+                        },
+                    ]}
+                    onSuccess={props.onSuccess}
+                    buttonLabel={"TOTP_CODE_CONTINUE_BUTTON"}
+                    callAPI={async (formFields) => {
+                        const totp = formFields.find((field) => field.id === "totp")?.value;
+                        if (totp === undefined || totp.length === 0) {
+                            throw new STGeneralError("GENERAL_ERROR_TOTP_UNDEFINED");
+                        }
+                        let response: Awaited<
+                            ReturnType<RecipeInterface["verifyCode"]> | ReturnType<RecipeInterface["verifyDevice"]>
+                        >;
+                        if (props.featureState.deviceInfo) {
+                            response = await props.recipeImplementation.verifyDevice({
+                                deviceName: props.featureState.deviceInfo!.deviceName,
+                                totp,
+                                userContext,
+                            });
+                        } else {
+                            response = await props.recipeImplementation.verifyCode({
+                                totp,
+                                userContext,
+                            });
+                        }
+
+                        // We can return these statuses, since they all cause a redirection or are handled elsewhere
+                        // so we don't really want to show anything
+                        if (
+                            response.status === "OK" ||
+                            response.status === "UNKNOWN_DEVICE_ERROR" ||
```

Comment:
How deos the Ui react to UNKNOWN_DEVICE_ERROR? This means that the device that was just created doesn't exist anymore, so does this ask the user to create a new device again?

Reply:
it shows the access denied screen telling the user to refresh the page.